File: benchmarks/cross-language/go/skynet.go
// Go Skynet Benchmark
// Based on https://github.com/atemerev/skynet
// Recursive tree of goroutines: root spawns 10 children, each spawns 10 more, etc.
// Leaves report their offset; parents aggregate 10 children's results and report up.
package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func getLeaves() int64 {
	if env := os.Getenv("SKYNET_LEAVES"); env != "" {
		if n, err := strconv.ParseInt(env, 10, 64); err == nil {
			return n
		}
	}
	if env := os.Getenv("BENCHMARK_MESSAGES"); env != "" {
		if n, err := strconv.ParseInt(env, 10, 64); err == nil {
			return n
		}
	}
	return 1000000
}

// skynetNode sends its subtree sum to the result channel.
// Leaves send their offset directly; internal nodes spawn 10 children and sum.
func skynetNode(result chan<- int64, offset, size int64) {
	if size == 1 {
		result <- offset
		return
	}
	children := make(chan int64, 10)
	childSize := size / 10
	for i := int64(0); i < 10; i++ {
		go skynetNode(children, offset+i*childSize, childSize)
	}
	var sum int64
	for i := 0; i < 10; i++ {
		sum += <-children
	}
	result <- sum
}

func main() {
	numLeaves := getLeaves()

	// Total actors = sum of nodes at each level
	totalActors := int64(0)
	n := numLeaves
	for n >= 1 {
		totalActors += n
		n /= 10
	}

	fmt.Println("=== Go Skynet Benchmark ===")
	fmt.Printf("Leaves: %d\n\n", numLeaves)

	root := make(chan int64, 1)
	start := time.Now()
	go skynetNode(root, 0, numLeaves)
	sum := <-root
	elapsed := time.Since(start)

	elapsedNs := elapsed.Nanoseconds()
	elapsedUs := elapsedNs / 1000

	fmt.Printf("Sum: %d\n", sum)
	if elapsedUs > 0 {
		nsPerMsg := elapsedNs / totalActors
		throughputM := totalActors / elapsedUs
		leftover := totalActors - (throughputM * elapsedUs)
		throughputFrac := (leftover * 100) / elapsedUs
		fmt.Printf("ns/msg:         %d\n", nsPerMsg)
		fmt.Printf("Throughput:     %d.", throughputM)
		if throughputFrac < 10 {
			fmt.Printf("0")
		}
		fmt.Printf("%d M msg/sec\n", throughputFrac)
	}
}

File: benchmarks/cross-language/erlang/skynet.erl
%% Erlang Skynet Benchmark
%% Based on https://github.com/atemerev/skynet
%% Uses native Erlang processes (lightweight ~2KB each).
%% For 1M leaves, start with: erl +P 2000000 -noshell -s skynet start
-module(skynet).
-export([start/0]).

get_leaves() ->
    case os:getenv("SKYNET_LEAVES") of
        false ->
            case os:getenv("BENCHMARK_MESSAGES") of
                false -> 1000000;
                Val -> list_to_integer(Val)
            end;
        Val -> list_to_integer(Val)
    end.

%% Compute total actors: sum of nodes at each level
total_actors(N) when N < 1 -> 0;
total_actors(N) -> N + total_actors(N div 10).

%% Each node spawns 10 children or reports its offset (leaf).
%% Results bubble up via message passing.
skynet_node(Offset, 1, Parent) ->
    Parent ! Offset;
skynet_node(Offset, Size, Parent) ->
    ChildSize = Size div 10,
    Self = self(),
    lists:foreach(fun(I) ->
        spawn(fun() ->
            skynet_node(Offset + I * ChildSize, ChildSize, Self)
        end)
    end, lists:seq(0, 9)),
    Sum = collect(10, 0),
    Parent ! Sum.

collect(0, Acc) -> Acc;
collect(N, Acc) ->
    receive
        Value ->
            collect(N - 1, Acc + Value)
    end.

start() ->
    NumLeaves = get_leaves(),
    TotalActors = total_actors(NumLeaves),

    io:format("=== Erlang Skynet Benchmark ===~n"),
    io:format("Leaves: ~p~n~n", [NumLeaves]),

    Self = self(),
    Start = erlang:monotonic_time(nanosecond),
    spawn(fun() -> skynet_node(0, NumLeaves, Self) end),
    Sum = receive V -> V after 120000 -> timeout end,
    End = erlang:monotonic_time(nanosecond),

    ElapsedNs = End - Start,
    ElapsedUs = ElapsedNs div 1000,

    io:format("Sum: ~p~n", [Sum]),
    if
        ElapsedUs > 0 ->
            NsPerMsg = ElapsedNs div TotalActors,
            ThroughputM = TotalActors div ElapsedUs,
            Leftover = TotalActors - (ThroughputM * ElapsedUs),
            ThroughputFrac = (Leftover * 100) div ElapsedUs,
            io:format("ns/msg:         ~p~n", [NsPerMsg]),
            FracStr = if ThroughputFrac < 10 -> io_lib:format("0~p", [ThroughputFrac]);
                         true -> io_lib:format("~p", [ThroughputFrac])
                      end,
            io:format("Throughput:     ~p.~s M msg/sec~n", [ThroughputM, FracStr]);
        true -> ok
    end,
    halt(0).
